### ISSUE TITLE
Color-code product group in Shopping list

### ISSUE THREAD

**Issue by regystro** [grocy/grocy#2164]
Use case:
To be able to color-code the shopping list items based on product group. That would help users to visually identify the shop section, approximate the amount of each group items, etc.

A few ideas on how it could be implemented:

- Add an entity "color" (could include a color picker) to the userfields type selector and export it to a CSS class so it can be used in `custom_css.html` while drawing the Shopping List.
- Simply export any specific user field (e.g. integer number) and have it be identified when using `custom_js.html` and `custom_css.html`. I guess I could manage to apply custom CSS style based on that integer field.

Thank you.

**Comment by regystro** [comment]
A third and I think an easier option:
- Add `data-product-group-id` to each product on the Shopping list

Now, looking at the rendered HTML, we have e.g. `data-product-id="3"` which allows me to track the product id. I'm only asking to add the product group identifier, so I can track it in custom_js.html and custom_css.html and apply color accordingly.

**Comment by berrnd** [comment]
Something which will require custom JS/CSS by default will never be implemented, since it's nothing for the average user.

If, then this will be simply another property "Color" of product groups which will then be used accordingly.

**Comment by regystro** [comment]
I agree custom JS/CSS is not for the average user, but wouldn't it be useful for users that like to customize CSS to be able to tell "which product group are we in"?

I already have a custom userfield to store the group color. Is it difficult to add `data-product-group-id` to each product on the Shopping list? As I suggested in my second post.

Thank you.

**Comment by berrnd** [comment]
I haven't said that this is not useful, otherwise I would have closed the request accordingly.

If you want to use custom JS/CSS anyway, I don't really get why exposing any id's or DOM attributes seems to be the key for you - the product group (name) and any Userfields is already rendered, simply use the value to do any styling. Technically not really different than using an `id` or any other attribute.

And of course nothing is technically really difficult or impossible, this here is not rocket science - implementing something just requires time.

**Comment by regystro** [comment]
Well, I was hoping to have some abstraction and rename/reassign capabilities using ids for product groups, but I'll use the name to identify the group in my custom CSS as you suggested.

Thanks for pointing me in the right direction and for your prompt responses :)

**Comment by regystro** [comment]
Hi again and sorry to bother you. After your kind suggestion, I've been testing my proof of concept, but I found that since there is no hierarchy between groups (those with data-name) and items (shoppinglistitem-*) I cannot find the way to apply the style.

Given this code tested in Chrome console:
```
$('*[id*=shoppinglistitem]:visible').each(function() {
    if ($(this).prev().attr("data-name")=="Vegetables") $(this).css("background-color","LightGreen");
    if ($(this).prev().attr("data-name")=="Meat") $(this).css("background-color","LightPink");
});
```
It only paints the first item:
![sl1](https://user-images.githubusercontent.com/5755481/224783138-28e1dc40-d671-41e3-baa0-08af53a28571.PNG)

Can you help me to achieve coloring all the subitems belonging to a group?

I guess either adding a `data-product-group-id` on every shoppinglistitem, or creating a hierarchy in the DOM where items are children of their group would be the key.

![sl2](https://user-images.githubusercontent.com/5755481/224783684-7aef1664-9d19-44d2-b9a3-392784a669c0.PNG)

Also beware, because I think there's a bug in the generated HTML. As you can see above, the `data-name` attribute for the Ungrouped contains HTML code `<span class="font-italic font-weight-light">Ungrouped</span>`

**Comment by berrnd** [comment]
grocy's web frontend uses [DataTables](https://datatables.net/) for displaying any table and grocy has [an API](https://demo-prerelease.grocy.info/api). So before fiddling around with raw DOM trees (which are also depended on which columns are shown, how they are ordered and so on) and insisting on adding new data attributes making this special use case more comfortable for your example JS snippet, I would simply use those managed ways to get the needed data and do any further stuff based on that.

Working commented example snippet (for `data/custom_js.html`) using the grocy API way:
```html
<script>
if (Grocy.ActiveNav == "shoppinglist")
{
	// Define product group colors
	const colors = {
		"01 Sweets": "red",
		"02 Bakery products": "yellow",
		"03 Tinned food": "green",
		"04 Butchery products": "blue",
		"05 Vegetables/Fruits": "purple",
		"06 Refrigerated products": "brown"
	}

	// Get all active products
	Grocy.Api.Get("objects/products?query[]=active=1",
		function(products)
		{
			// Get all product groups
			Grocy.Api.Get("objects/product_groups",
				function(productGroups)
				{
					// Get all items of the currently displayed shopping list
					Grocy.Api.Get("objects/shopping_list?query[]=shopping_list_id=" + $("#selected-shopping-list").val(),
						function(shoppingListItems)
						{
							// Loop through all shopping list items
							$.each(shoppingListItems, function()
							{
								if (!this.product_id)
								{
									// Not a product shopping list item...
									return;
								}

								// Get the corresponding product
								var product = FindObjectInArrayByPropertyValue(products, "id", this.product_id);

								if (!product.product_group_id)
								{
									// Product has no product group...
									return;
								}

								// Get the corresponding product group
								var productGroup = FindObjectInArrayByPropertyValue(productGroups, "id", product.product_group_id);

								// Set the color of the corresponding table row (cells)
								$("#shoppinglistitem-" + this.id + "-row td").css("background-color", colors[productGroup.name]);
							});
						}
					);
				}
			);
		}
	);
}
</script>
```

&nbsp;
> Also beware, because I think there's a bug in the generated HTML.

`data-` attributes can contain arbitrary text and HTML is just text. If you think this shouldn't be allowed for whatever reason, I guess you need to take care changing the corresponding specification.

**Comment by regystro** [comment]
I just tested the code and it works flawlessly. Thank you very much!!

I would suggest to include those very useful code examples in the documentation (maybe near https://github.com/grocy/grocy#adding-your-own-css-or-js-without-to-have-to-modify-the-application-itself) and also the functions available in https://github.com/grocy/grocy/blob/master/public/js/grocy.js, so people could use it right away. Right now searching for Grocy.Api.Get in issues only gives 3 results (including this one) and I think is hard to find. Sorry, this is my first time dealing with API in javascript.

Thank you again and greetings!!
